Add CI for Windows

CarterLi opened the pull request:
for both AMD64 and ARM64

![image](https://github.com/user-attachments/assets/262003fe-b3ff-4a11-bddc-46de6c54d8b8)

ludouzi commented:
Thanks for all the work you've put into this. I'll find some time to give it a full review later, and after I can test it in Windows.

After a quick glance, only issues I have with the C++ are minor style inconsistencies. For the CMake, `INSTALL_FHS` needs to be turned off for Windows so everything builds into a single directory.

CarterLi commented:
Rebased. Added copyright headers and fixed several style inconsistencies.

> For the CMake, `INSTALL_FHS` needs to be turned off for Windows so everything builds into a single directory.

We have a lot of DLL dependencies. I plan to submit fooyin to [MSYS2](https://github.com/msys2/MINGW-packages) which uses pacman and manages dependencies automatically. MSYS2 does use FHS.

Incomplete dependencies found by [System Informer](https://systeminformer.com/)

```
fooyin.exe, 0x7ff711ca0000, 412 kB, A customisable music player
advapi32.dll, 0x7ffff4780000, 720 kB, 高级 Windows 32 基本 API
libc++.dll, 0x7fff25030000, 1.33 MB, 
libfooyin_core.dll, 0x7fff3c310000, 1.62 MB, 
avcodec-61.dll, 0x7ffeefd90000, 22.01 MB, FFmpeg codec library
libaom.dll, 0x7ffeec160000, 9.15 MB, 
libcairo-2.dll, 0x7ffeed4c0000, 1.16 MB, 
libfontconfig-1.dll, 0x7fff01670000, 344 kB, 
libexpat-1.dll, 0x7fff20430000, 204 kB, 
libpixman-1-0.dll, 0x7ffeea4f0000, 856 kB, 
msimg32.dll, 0x7fffec4f0000, 32 kB, GDIEXT Client DLL
libdav1d-7.dll, 0x7ffeee3e0000, 2.07 MB, dav1d 1.5.1 - AV1 decoder
libglib-2.0-0.dll, 0x7fff18280000, 1.41 MB, GLib
libpcre2-8-0.dll, 0x7ffeea5d0000, 756 kB, 
libgobject-2.0-0.dll, 0x7fff8f870000, 344 kB, GObject
libffi-8.dll, 0x7fffe9bd0000, 56 kB, 
libgsm.dll, 0x7fffb9390000, 88 kB, 
libiconv-2.dll, 0x7fff1dab0000, 1.08 MB, LGPLed libiconv for Windows
libintl-8.dll, 0x7fffb7170000, 312 kB, LGPLed libintl for Windows
libjxl.dll, 0x7ffeebd40000, 4.08 MB, 
libbrotlienc.dll, 0x7ffeea430000, 736 kB, 
libhwy.dll, 0x7fff6c000000, 68 kB, 
libjxl_cms.dll, 0x7fff55a40000, 140 kB, 
liblcms2-2.dll, 0x7ffee6c20000, 464 kB, 
libjxl_threads.dll, 0x7fffeac90000, 48 kB, 
liblc3-1.dll, 0x7fffa12a0000, 200 kB, 
liblzma-5.dll, 0x7fffb85c0000, 208 kB, liblzma data compression library
libmp3lame-0.dll, 0x7fff20e90000, 560 kB, 
libopencore-amrnb-0.dll, 0x7fff9b400000, 216 kB, 
libopencore-amrwb-0.dll, 0x7fffb86b0000, 144 kB, 
libopenjp2-7.dll, 0x7fff3bfe0000, 440 kB, 
libopus-0.dll, 0x7fff1b530000, 536 kB, 
librav1e.dll, 0x7ffeeba10000, 3.13 MB, 
librsvg-2-2.dll, 0x7ffeedf30000, 4.66 MB, 
bcryptprimitives.dll, 0x7ffff1b20000, 612 kB, Windows Cryptographic Primitives Library
libgdk_pixbuf-2.0-0.dll, 0x7fff22df0000, 340 kB, GIMP Toolkit
GdiPlus.dll, 0x7fffe8410000, 1.84 MB, Microsoft GDI+
libjpeg-8.dll, 0x7ffeea0a0000, 864 kB, 
libtiff-6.dll, 0x7ffeea180000, 600 kB, 
libdeflate.dll, 0x7fff25010000, 128 kB, 
libjbig-0.dll, 0x7fff39ea0000, 88 kB, 
libLerc.dll, 0x7ffee6b60000, 736 kB, 
libgio-2.0-0.dll, 0x7ffeea700000, 1.72 MB, Gio
libgmodule-2.0-0.dll, 0x7fffe8140000, 48 kB, GModule
shlwapi.dll, 0x7ffff3340000, 396 kB, 外壳简易实用工具库
libpango-1.0-0.dll, 0x7ffeea690000, 404 kB, Pango
libfribidi-0.dll, 0x7fff3b920000, 164 kB, 
libthai-0.dll, 0x7fff3bed0000, 88 kB, 
libdatrie-1.dll, 0x7fffe77d0000, 60 kB, 
libpangocairo-1.0-0.dll, 0x7fff6ccf0000, 84 kB, PangoCairo
libpangoft2-1.0-0.dll, 0x7fff60290000, 116 kB, PangoFT2
libpangowin32-1.0-0.dll, 0x7fff5f1e0000, 116 kB, PangoWin32
libunwind.dll, 0x7fff8e530000, 72 kB, 
libspeex-1.dll, 0x7fff960b0000, 208 kB, 
libSvtAv1Enc-3.dll, 0x7ffeeb160000, 8.66 MB, 
libtheoradec-2.dll, 0x7fff993d0000, 104 kB, 
libtheoraenc-2.dll, 0x7fff8e550000, 232 kB, 
libogg-0.dll, 0x7fffe9150000, 56 kB, 
libva.dll, 0x7fff3c2c0000, 280 kB, 
libvorbis-0.dll, 0x7fff601c0000, 248 kB, 
libvorbisenc-2.dll, 0x7fff19520000, 576 kB, 
libvpl-2.dll, 0x7fff214b0000, 420 kB, Intel® VPL Dispatcher
libvpx-1.dll, 0x7ffeee600000, 3.4 MB, 
libwinpthread-1.dll, 0x7fff8e5c0000, 96 kB, POSIX WinThreads for Windows
libwebp-7.dll, 0x7ffeed2d0000, 652 kB, 
libsharpyuv-0.dll, 0x7fffe8780000, 64 kB, 
libwebpmux-3.dll, 0x7fffd4ac0000, 92 kB, 
libx264-164.dll, 0x7ffeeaaa0000, 2.98 MB, H.264 (MPEG-4 AVC) encoder library
libx265-215.dll, 0x7ffee73c0000, 18.97 MB, HEVC video encoder
libzvbi-0.dll, 0x7fff67850000, 456 kB, 
xvidcore.dll, 0x7ffeea950000, 1.3 MB, 
avformat-61.dll, 0x7fff02450000, 2.42 MB, FFmpeg container format library
libbluray-2.dll, 0x7fff55060000, 376 kB, 
libbz2-1.dll, 0x7fff9afb0000, 112 kB, 
libgme.dll, 0x7fff02390000, 752 kB, 
libgnutls-30.dll, 0x7ffeeaf50000, 2.01 MB, 
crypt32.dll, 0x7ffff2280000, 1.46 MB, 加密 API32
cryptsp.dll, 0x7ffff1240000, 108 kB, Cryptographic Service Provider API
libidn2-0.dll, 0x7fff55970000, 248 kB, 
libp11-kit-0.dll, 0x7ffeea2d0000, 1.33 MB, 
libtasn1-6.dll, 0x7fff67830000, 120 kB, 
libunistring-5.dll, 0x7ffee7170000, 2.27 MB, LGPLed libunistring for Windows
ncrypt.dll, 0x7ffff1460000, 192 kB, Windows NCrypt 路由器
ntasn1.dll, 0x7ffff1410000, 252 kB, Microsoft ASN.1 API
libmodplug-1.dll, 0x7ffeede40000, 940 kB, 
librtmp-1.dll, 0x7fff98f50000, 128 kB, 
libgmp-10.dll, 0x7ffeea220000, 688 kB, 
libhogweed-6.dll, 0x7fff33dc0000, 284 kB, 
libnettle-8.dll, 0x7ffeeae30000, 340 kB, 
libsrt.dll, 0x7ffeeae90000, 716 kB, 
libcrypto-3-x64.dll, 0x7ffee6ca0000, 4.8 MB, OpenSSL library
wsock32.dll, 0x7fffe8750000, 40 kB, Windows Socket 32-Bit DLL
libssh.dll, 0x7fff3bd30000, 480 kB, 
libxml2-2.dll, 0x7ffeed380000, 1.24 MB, 
avutil-59.dll, 0x7ffeeec70000, 17.12 MB, FFmpeg utility library
bcrypt.dll, 0x7ffff1a10000, 152 kB, Windows 加密基元库
libva_win32.dll, 0x7fffe9c00000, 36 kB, 
libfooyin_utils.dll, 0x7fff993f0000, 480 kB, 
libtag-2.dll, 0x7fff98f70000, 880 kB, 
Qt6Network.dll, 0x7fff1e150000, 1.83 MB, C++ Application Development Framework
dnsapi.dll, 0x7ffff03f0000, 1.15 MB, DNS 客户端 API DLL
IPHLPAPI.DLL, 0x7ffff03b0000, 208 kB, IP 帮助程序 API
nsi.dll, 0x7ffff4420000, 40 kB, NSI User-mode interface DLL
libbrotlidec.dll, 0x7fffd4ae0000, 76 kB, 
libbrotlicommon.dll, 0x7fff5d3d0000, 160 kB, 
secur32.dll, 0x7fffeeaa0000, 52 kB, Security Support Provider Interface
sspicli.dll, 0x7ffff0cc0000, 292 kB, Security Support Provider Interface
winhttp.dll, 0x7fffeb2b0000, 1.12 MB, Windows HTTP 服务
Qt6Sql.dll, 0x7fffdf880000, 296 kB, C++ Application Development Framework
swresample-5.dll, 0x7fffebf70000, 164 kB, FFmpeg audio resampling library
libsoxr.dll, 0x7ffeea8c0000, 516 kB, 
libfooyin_gui.dll, 0x7fff1b5c0000, 3.59 MB, 
Qt6Concurrent.dll, 0x7fffebf60000, 56 kB, C++ Application Development Framework
Qt6Core.dll, 0x7fff183f0000, 5.66 MB, C++ Application Development Framework
authz.dll, 0x7ffff0570000, 320 kB, Authorization Framework
libb2-1.dll, 0x7fffeae90000, 64 kB, 
libdouble-conversion.dll, 0x7fffdb750000, 88 kB, 
libicuin77.dll, 0x7ffeee970000, 2.98 MB, 
libicuuc77.dll, 0x7fff016d0000, 1.79 MB, 
libicudt77.dll, 0x7ffee10c0000, 30.45 MB, 
libpcre2-16-0.dll, 0x7fff5c4d0000, 692 kB, 
libzstd.dll, 0x7fff204b0000, 984 kB, 
mpr.dll, 0x7fffdf850000, 132 kB, 多提供程序路由器 DLL
netapi32.dll, 0x7fffd6340000, 104 kB, Net Win32 API DLL
netutils.dll, 0x7ffff03a0000, 52 kB, Net Win32 API Helpers DLL
srvcli.dll, 0x7fffdecf0000, 164 kB, Server Service Client DLL
userenv.dll, 0x7ffff1040000, 172 kB, Userenv
version.dll, 0x7fffea510000, 44 kB, Version Checking and File Installation Libraries
winmm.dll, 0x7fffe5450000, 212 kB, MCI API DLL
ws2_32.dll, 0x7ffff46b0000, 464 kB, Windows Socket 2.0 32 位 DLL
Qt6Gui.dll, 0x7fff026c0000, 8.51 MB, C++ Application Development Framework
d3d11.dll, 0x7fffede50000, 2.38 MB, Direct3D 11 Runtime
D3D12.dll, 0x7fff6bf50000, 140 kB, Direct3D 12 Runtime
DWrite.dll, 0x7fffed5a0000, 2.4 MB, Microsoft DirectX 版式服务
dxgi.dll, 0x7fffeef10000, 1.21 MB, DirectX Graphics Infrastructure
libfreetype-6.dll, 0x7fff21520000, 812 kB, Font Rendering Library
libharfbuzz-0.dll, 0x7fff38740000, 1.2 MB, 
libgraphite2.dll, 0x7fff96170000, 152 kB, 
usp10.dll, 0x7fffe8730000, 100 kB, Uniscribe Unicode script processor
libmd4c.dll, 0x7fffe7360000, 104 kB, 
libpng16-16.dll, 0x7fffd48a0000, 260 kB, 
ole32.dll, 0x7ffff3f50000, 1.59 MB, 用于 Windows 的 Microsoft OLE
zlib1.dll, 0x7fffe7340000, 124 kB, 
Qt6Widgets.dll, 0x7fff195b0000, 6.48 MB, C++ Application Development Framework
dwmapi.dll, 0x7fffef070000, 184 kB, Microsoft 桌面窗口管理器 API
uxtheme.dll, 0x7fffeed80000, 676 kB, Microsoft UxTheme 库
shell32.dll, 0x7ffff2750000, 7.27 MB, Windows Shell 公用 DLL
ucrtbase.dll, 0x7ffff2100000, 1.29 MB, Microsoft® C Runtime Library
```

ludouzi commented:
For official Windows build support, we will be using `vcpkg` for ease of dependency management. The CI (in the [windows](https://github.com/fooyin/fooyin/tree/windows) branch) now produces both a zip archive and an installer, which will be used for future releases once merged.

If you prefer MSYS2, you are of course welcome to maintain a separate version, but official releases will focus on vcpkg for now.

The two plugins you’ve added look good. To keep things clean, it may be best to submit them as separate PRs.
